core/data/tools_cli.joke

```diff
@@ -0,0 +1,706 @@
+      (validate value spec opt optarg))))
+
+(defn- neg-flag? [spec opt]
+  (and (:long-opt spec)
+       (re-find #"^--\[no-\]" (:long-opt spec))
+       (re-find #"^--no-" opt)))
+
+(defn- parse-optarg [spec opt optarg]
+  (let [{:keys [required]} spec]
+    (if (and required (nil? optarg))
+      [::error (missing-required-error opt required)]
+      (let [value (if required
+                    optarg
+                    (not (neg-flag? spec opt)))]
+        (parse-value value spec opt optarg)))))
+
+(defn- parse-option-tokens
+  "Reduce sequence of [opt-type opt ?optarg?] tokens into a map of
+  {option-id value} merged over the default values in the option
+  specifications.
+
+  If the :no-defaults flag is true, only options specified in the tokens are
+  included in the option-map.
+
+  Unknown options, missing options, missing required arguments, option
+  argument parsing exceptions, and validation failures are collected into
+  a vector of error message strings.
+
+  If the :strict flag is true, required arguments that match other options
+  are treated as missing, instead of a literal value beginning with - or --.
+
+  Returns [option-map error-messages-vector]."
+  [specs tokens & options]
+  (let [{:keys [no-defaults strict]} (apply hash-map options)
+        defaults (default-option-map specs :default)
+        default-fns (default-option-map specs :default-fn)
+        requireds (missing-errors specs)]
+    (-> (reduce
+          (fn [[m ids errors] [opt-type opt optarg]]
+            (if-let [spec (find-spec specs opt-type opt)]
+              (let [[value error] (parse-optarg spec opt optarg)
+                    id (:id spec)]
+                (if-not (= value ::error)
+                  (if (and strict
+                           (or (find-spec specs :short-opt optarg)
+                               (find-spec specs :long-opt optarg)))
+                    [m ids (conj errors (missing-required-error opt (:required spec)))]
+                    [(if-let [update-fn (:update-fn spec)]
+                       (update-in m [id] update-fn)
+                       ((:assoc-fn spec assoc) m id value))
+                     (conj ids id)
+                     errors])
+                  [m ids (conj errors error)]))
+              [m ids (conj errors (str "Unknown option: " (pr-str opt)))]))
+          [defaults [] []] tokens)
+        (#(reduce
+           (fn [[m ids errors] [id error]]
+             (if (contains? m id)
+               [m ids errors]
+               [m ids (conj errors error)]))
+           % requireds))
+        (#(reduce
+           (fn [[m ids errors] [id f]]
+             (if (contains? (set ids) id)
+               [m ids errors]
+               [(assoc m id (f (first %))) ids errors]))
+           % default-fns))
+        (#(let [[m ids errors] %]
+            (if no-defaults
+              [(select-keys m ids) errors]
+              [m errors]))))))
+
+(defn ^{:added "0.3.0"} make-summary-part
```

Comment:
All `:added` tags need to be changed to `^{:added "1.0"}`. I know they are meaningless until v1.0 actually comes out, but that's the convention for pre-v1.0 API.